### PR TITLE
✨[amp-form] allow `form` attributes for form elements outside of `amp-form`

### LINKED ISSUE #32447: Do not restrict the "form" attribute on inputs inside a form tag
Per - https://amp.dev/documentation/components/amp-form/#not-allowed - the form attribute seems to be restricted
```
<input type=button>, <input type=image> - Most of the form-related attributes on inputs including: form, formaction, formtarget, formmethod and others.

(Relaxing some of these rules might be reconsidered in the future - please let us know if you require these and provide use cases).
```

Usecase

if i have inputs on my form on desktop and use amp-sidebar's toolbar to put it elsewhere on the page at different breakpoints, 
ill need the form to consider these as it's own inputs at all breakpoints. 

(in my site , i am doing some massive filters and they are under one form on desktop and split across 3 sidebars on mobile , they use the toolbar feature to make the inputs and all functioning mechanisms non-redundant)

I think enabling the `form` attribute is not going to cause any content shifts (since developers normally dont target it in their CSS)

### PR DESCRIPTION
**Pull request by micajuine-ho**
Closes #32447

This PR removes the restriction on amp-forms to only have form elements within the form. Notably, this means that input elements can be outside of the form when `form='form_id'`. Wherever we performed operations on form elements that matched certain selectors (i.e. for sanitizing, variable substitution, validity checking, etc), we now don't constrain those elements to be contained by the form element.

Additionally, this PR moves the event handling from the amp-form element to the AmpFormService. The single AmpFormService, is in charge setting up and disposing of listeners on the document. The listeners listen for the form events and the AmpFormService delegates the event to the correct AmpForm element. 

TODO: Verify amp-selector in amp form does not allow this behavior.

Changes inclue:
- Validator changes
- Unit tests added
- Existing Unit tests fixed 
- Allowlist of `getServiceForDocOrNull` (AmpFormService should always be available before the element is created (along with the DirtinessHandler))
- Removed constraint documentation

### CODE REVIEW

File: validator/validator-main.protoascii
Comment (honeybadgerdontcare) [inline validator/validator-main.protoascii:2741]: Please add a `descriptive_name` as well. This is what would be used in the error messages shown to developers.

https://github.com/ampproject/amphtml/blob/master/validator/validator.proto#L569

File: validator/testdata/feature_tests/forms.out
Comment (honeybadgerdontcare) [inline validator/testdata/feature_tests/forms.out:136]: Does this need to be updated with `gulp validator --update-tests` still?

**Review by honeybadgerdontcare — CHANGES_REQUESTED** [review]

File: validator/validator-main.protoascii
Comment (honeybadgerdontcare) [inline validator/validator-main.protoascii:2740]: is there a reason to create a new tag spec? why not add `form` as an attribute to `input-common-attrs`?

https://github.com/ampproject/amphtml/blob/master/validator/validator-main.protoascii#L2667

This is probably why the test is failing. It's testing against the first input tagspec.

File: extensions/amp-form/0.1/amp-form.js
Comment (alanorozco) [inline extensions/amp-form/0.1/amp-form.js:1469]: This would be simpler if we try to align with the signature for `querySelector` so we can replace other uses wholesale.

```js
function querySelectorAllFormElements(form, selector) {}
```

It selects those inside `form`, and `form.elements` filtered with [`matches()`](https://developer.mozilla.org/en-US/docs/Web/API/Element/matches).

File: extensions/amp-form/0.1/test/test-amp-form.js
Comment (alanorozco) [inline extensions/amp-form/0.1/test/test-amp-form.js:2386]: There are more featuers to test. A way to ensure that we run each tests in two forms would be ideal:

(super abstract, doesn't need this signature)

```js
const inside = html`
  <form>
    <input />
  </form>
`;
const outside = html`
  <div>
    <input form="foo" />
    <form id="foo">...</form>
  </div>
`;
myTestCase1(inside, inside.querySelector('input'));
myTestCase2(outside, inside.querySelector('input'));
```

I also think we should try to keep the structure of existing tests as much as possible, and reuse them in a way that makes it clear that we're testing the same behavior in two diff tree modes.

**Review by alanorozco — CHANGES_REQUESTED** [review]
There's several more features that will not work. Look for other `querySelector(All)`s on the page.

**Review by alanorozco — COMMENTED** [review]
Don't forget to re-scope these event handlers as well.

https://github.com/ampproject/amphtml/blob/feb6466190b2e12971af1b558e9dc3fcc33d6c13/extensions/amp-form/0.1/amp-form.js#L409-L476

They should listen to `Document` instead and dispatch their handler if they belong to this form (hint: target elements have `element.form`). 

Since we're also now listening to Document events, event handling becomes slower since all forms will listen to all events. I suggest you centralize these handlers so that you only install one event type handler per Document, which would get dispatched to the correct form.

**Comment by alanorozco** [comment]
One more thing to look for, is that all our documented cases should work in both tree modes. Manual testing gets you far to ensure that you're covering all features, but e2e testing will be key here.

File: validator/validator-main.protoascii
Comment (honeybadgerdontcare) [reply validator/validator-main.protoascii:2741]: Resolving as I believe there shouldn't be a new tagspec for this but included in input-common-attr attribute list.

File: validator/validator-main.protoascii
Comment (honeybadgerdontcare) [reply validator/validator-main.protoascii:2740]: Resolving as I believe there shouldn't be a new tagspec for this but included in input-common-attr attribute list.

File: validator/testdata/feature_tests/forms.out
Comment (honeybadgerdontcare) [reply validator/testdata/feature_tests/forms.out:136]: Resolving as I believe there shouldn't be a new tagspec for this but included in input-common-attr attribute list.

**Comment by honeybadgerdontcare** [comment]
Please consider having the `form` attribute as a part of the `input-common-attrs`. This would permit it on the current variations of `input` tag. Since those are available for AMP Ads and AMP Email as well you'll want to get their go-ahead for that. If they do not want to do it then we could add a `disabled_by: "amp4email"` on the attribute in the attribute list.

See cl/361250383 for implementation details.

**Comment by micajuine-ho** [comment]
@honeybadgerdontcare  SG, will make the appropriate changes. Thanks for the reviews.

File: extensions/amp-form/0.1/amp-form.js
Comment (alanorozco) [inline extensions/amp-form/0.1/amp-form.js:430]: Is there a type of extension service getter that is not async? (Since the extension is loaded anyway...)

File: extensions/amp-form/0.1/amp-form.js
Comment (alanorozco) [inline extensions/amp-form/0.1/amp-form.js:1487]: I'm not sure it has to be `INPUT`. There's `textarea`, etc.  Why this condition?

```suggestion
```

**Review by alanorozco — COMMENTED** [review]
The approach looks good!

Just my one question about the async service getter, then I can look at the remaining details for approval.

File: extensions/amp-form/0.1/amp-form.js
Comment (micajuine-ho) [reply extensions/amp-form/0.1/amp-form.js:430]: Is it guaranteed that the service loads before the AmpForm class?

File: extensions/amp-form/0.1/amp-form.js
Comment (micajuine-ho) [reply extensions/amp-form/0.1/amp-form.js:1487]: Are other `form` input elements (i.e. `textarea`), allowed to be outside of the form? I thought that wasn't the case, and that this PR is changing that too allow `inputs` to be outside now.

File: extensions/amp-form/0.1/amp-form.js
Comment (micajuine-ho) [reply extensions/amp-form/0.1/amp-form.js]: Seems like the sync version is preferred: 
`148:28  error  Forbidden: "getServiceForDoc". Synchronous access to element services is unreliable (#22414). Use getServicePromiseForDoc() instead  local/no-forbidden-terms`

See https://github.com/ampproject/amphtml/issues/22414

I feel that sync should be ok, but then again it could cause weird issues that are hard to repro/track down in the future, all for a small performance benefit. What are your thoughts?

File: extensions/amp-form/0.1/amp-form.js
Comment (alanorozco) [inline extensions/amp-form/0.1/amp-form.js:429]: Short-circuit instead. Makes the diff cleaner and the code more shallow.

```suggestion
    if (!this.ampFormService_) {
      return;
    }
```

But it should absolutely be available, so even better:

```suggestion
    devAssert(this.ampFormService_);
```

File: extensions/amp-form/0.1/amp-form.js
Comment (alanorozco) [reply extensions/amp-form/0.1/amp-form.js:1486]: Yep, all native input elements should be supported. This includes `input`, `textarea`, `select` and maybe others I'm forgetting.

I believe the `.form`/`.elements` relationship is all you need to confirm, so removing the check as per my earlier suggestion should be the way to go.

File: extensions/amp-form/0.1/amp-form.js
Comment (alanorozco) [inline extensions/amp-form/0.1/amp-form.js:1483]: nit: implicit return. Including my other suggestion below:

```suggestion
  return Array.from(form.elements).filter((element) => matches(element, query));
```

File: extensions/amp-form/0.1/amp-form.js
Comment (alanorozco) [inline extensions/amp-form/0.1/amp-form.js:1752]:
```suggestion
            // Only call handler if the element has a registered form.
```

File: extensions/amp-form/0.1/form-verifiers.js
Comment (alanorozco) [inline extensions/amp-form/0.1/form-verifiers.js:252]: nit: maybe extract `formElementsQuerySelector`. This case may not be problematic since the selector is likely unique, but I'd prefer if we avoid creating and slicing arrays for a single item.


```js
function formElementsQuerySelector(form, query) {
  for (let i = 0; i < form.elements.length; i++) {
    const element = form.elements[i];
    if (matches(element, query)) {
      return element;
    }
  }
  return null;
}
```

```suggestion
        (e) => formElementsQuerySelector(this.form_, `[name="${e.name}"]`)
```

**Review by alanorozco — COMMENTED** [review]
We're close! Mainly the problem of filtering by tagname, plus a couple of nits.

**Comment by Gregable** [comment]
Adding myself to reviewers while @honeybadgerdontcare  is out. Looks like his concern is so far unresolved. I'm happy to approve after.

File: extensions/amp-form/0.1/amp-form.js
Comment (micajuine-ho) [reply extensions/amp-form/0.1/amp-form.js]: Discussed offline: we should be good.

**Review by alanorozco — CHANGES_REQUESTED** [review]
This has been a hard change!  Thank you for doing so much work so that we can allow the `form` attribute correctly.

I realized we're breaking support for events from descendant AMP elements like `<amp-selector>`. This should be a matter of applying my suggested fix.

However, we may need to work closer on ensuring a good testing surface. We're syncing offline, but I'm happy to provide help with testing.

File: extensions/amp-form/0.1/amp-form.js
Comment (micajuine-ho) [reply extensions/amp-form/0.1/amp-form.js:400]: Great write up. Thanks!

File: extensions/amp-form/0.1/amp-form.js
Comment (micajuine-ho) [reply extensions/amp-form/0.1/amp-form.js:1735]: Should we also cache the form per target? Could save us a good amount of small traversals, especially for events such as `AMP_EVENTS.FORM_VALUE_CHANGE`

**Comment by micajuine-ho** [comment]
@alanorozco PTAL

**Comment by micajuine-ho** [comment]
@alanorozco PTAL when you have a chance.

File: extensions/amp-form/0.1/amp-form.js
Comment (alanorozco) [inline extensions/amp-form/0.1/amp-form.js]:
```suggestion
              dev().assertElement(e.target);
```

**Review by alanorozco — APPROVED** [review]
LGTM per the dev.assertElement change

**Review by honeybadgerdontcare — APPROVED** [review]
lgtm for validation, thank you for making the recommended changes.

**Review by rcebulko — APPROVED** [review]
forbidden-terms

File: build-system/test-configs/forbidden-terms.js
Comment (rsimha) [inline build-system/test-configs/forbidden-terms.js:310]: nit: As written, this comment looks like it applies to lines 309-322 when it likely only applies to lines 309-310. Here's a suggested alternative for better readability.
```suggestion
      'extensions/amp-form/0.1/amp-form.js', // References service defined in amp-form.
      'extensions/amp-form/0.1/form-dirtiness.js', // References service defined in amp-form.
```

**Review by rsimha — APPROVED** [review]

**Comment by zhangsu** [comment]
As mentioned by https://github.com/ampproject/amphtml/pull/33095#issuecomment-791815889 for spec changes affecting email please give up a heads up so we can review the change presubmit (as opposed to reviewing it post-mortem and having to block our releases because of unintended changes). In this case the change looks fine but it would block our releases for various email providers otherwise.

You can notify the email working group by tagging @ampproject/wg-amp4email

**Comment by micajuine-ho** [comment]
@zhangsu Sorry about that, it slipped through the cracks. I'll make sure to notify beforehand in future changes.

**Pull request merged**
